apps/els_lsp/src/els_code_action_provider.erl

```diff
@@ -29,79 +29,133 @@ handle_request({document_codeaction, Params}, State) ->
+       fun action_fix_module_name/3}
+    ], Uri, Range, Message).
+
+-spec make_code_action([{string(), Fun}], uri(), range(), binary()) -> [map()]
+          when Fun :: fun((uri(), range(), [binary()]) -> [map()]).
+make_code_action([], _Uri, _Range, _Message) ->
+  [];
+make_code_action([{RE, Fun}|Rest], Uri, Range, Message) ->
+  Actions = case re:run(Message, RE, [{capture, all_but_first, binary}]) of
+              {match, Matches} ->
+                Fun(Uri, Range, Matches);
+              nomatch ->
+                []
+            end,
+  Actions ++ make_code_action(Rest, Uri, Range, Message).
+
+-spec action_export_function(uri(), range(), [binary()]) -> [map()].
+action_export_function(Uri, _Range, [UnusedFun]) ->
+  {ok, Document} = els_utils:lookup_document(Uri),
+  case els_poi:sort(els_dt_document:pois(Document, [module, export])) of
+    [] ->
+      [];
+    POIs ->
+      #{range := #{to := {Line, _Col}}} = lists:last(POIs),
+      Pos = {Line + 1, 1},
+      [ make_edit_action( Uri
+                        , <<"Export ", UnusedFun/binary>>
+                        , ?CODE_ACTION_KIND_QUICKFIX
+                        , <<"-export([", UnusedFun/binary, "]).\n">>
+                        , els_protocol:range(#{from => Pos, to => Pos})) ]
+  end.
+
+-spec action_ignore_variable(uri(), range(), [binary()]) -> [map()].
+action_ignore_variable(Uri, Range, [UnusedVariable]) ->
+  {ok, Document} = els_utils:lookup_document(Uri),
+  POIs = els_poi:sort(els_dt_document:pois(Document, [variable])),
+  case ensure_range(els_range:to_poi_range(Range), UnusedVariable, POIs) of
+    {ok, VarRange} ->
+      [ make_edit_action( Uri
+                        , <<"Add '_' to '", UnusedVariable/binary, "'">>
+                        , ?CODE_ACTION_KIND_QUICKFIX
+                        , <<"_", UnusedVariable/binary>>
+                        , els_protocol:range(VarRange)) ];
+    error ->
+      []
+  end.
+
+-spec action_suggest_variable(uri(), range(), [binary()]) -> [map()].
+action_suggest_variable(Uri, Range, [Var]) ->
+  %% Supply a quickfix to replace an unbound variable with the most similar
+  %% variable name in scope.
+  {ok, Document} = els_utils:lookup_document(Uri),
+  POIs = els_poi:sort(els_dt_document:pois(Document, [variable])),
+  case ensure_range(els_range:to_poi_range(Range), Var, POIs) of
+    {ok, VarRange} ->
+      ScopeRange = els_scope:variable_scope_range(VarRange, Document),
+      VarsInScope = [atom_to_binary(Id, utf8) ||
+                      #{range := R, id := Id} <- POIs,
+                      els_range:in(R, ScopeRange),
+                      els_range:compare(R, VarRange)],
+      case [{els_utils:levenshtein_distance(V, Var), V} ||
+             V <- VarsInScope,
+             V =/= Var,
+             binary:at(Var, 0) =:= binary:at(V, 0)]
+      of
+        [] ->
+          [];
+        VariableDistances ->
+          {_, SimilarVariable} = lists:min(VariableDistances),
```

Comment:
Maybe we should also set a threshold as well, so that completely different variable names are not presented as an option.